src/Workspaces/CSharp/Portable/Editing/CSharpImportAdder.cs

```diff
@@ -66,138 +66,152 @@ protected override SyntaxNode MakeSafeToAddNamespaces(SyntaxNode root, IEnumerab
             return null;
         }
 
-        private class Rewriter : CSharpSyntaxRewriter
+        /// <summary>
+        /// Walks the portion of the tree we're adding imports to looking to see if those imports could likely cause
+        /// conflicts with existing code.  Note: this is a best-effort basis, and the goal is to catch reasonable
+        /// conflicts effectively.  There may be cases that do slip through that we can adjust for in the future.  Those
+        /// cases should be assessed to see how reasonable/likely they are.  I.e. if it's just a hypothetical case with
+        /// no users being hit, then that's far less important than if we have a reasonable coding pattern that would be
+        /// impacted by adding an import to a normal namespace.
+        /// </summary>
+        private class ConflictWalker : CSharpSyntaxWalker
         {
             private readonly SemanticModel _model;
-            private readonly Workspace _workspace;
             private readonly CancellationToken _cancellationToken;
 
             /// <summary>
-            /// A hashset containing the short names of all namespace members 
+            /// A mapping containing the simple names and arity of all namespace members, mapped to the import that
+            /// they're brought in by.
             /// </summary>
-            private readonly HashSet<string> _namespaceMembers;
+            private readonly MultiDictionary<(string name, int arity), INamespaceSymbol> _namespaceMembers
+                = new MultiDictionary<(string name, int arity), INamespaceSymbol>();
+
+            /// <summary>
+            /// A mapping containing the simple names of all extension methods, mapped to the import that they're
+            /// brought in by.  This doesn't keep track of arity because methods can be called with type arguments.
+            /// </summary>
```

Comment:
You don't have arity, but was there any thought about parameter count? There's unfortunately a lot of tricky params/optional cases that make it a bit tricky to get right, but was there any concern about this being a bit too overly broad?

Reply:
> There's unfortunately a lot of tricky params/optional cases that make it a bit tricky to get right, 

YEah... overload resolution is just insane (including just thinking about param count).  I really didn't want to go to that level.  We can see if this is overbroad and we can lighten later.  However, i don't mind the restriction for now.

Reply:
doc'ed this